### DISCUSSION TITLE
Public role permission on files in subfolders

### DISCUSSION THREAD

**Discussion by keesvanbemmel** [directus/directus#14353] [Archive]
Hi!

What is the most efficient filter to set when I want the public role to access files in a folder and all its subfolders?

Cheers!
Kees.

**Comment by rijkvanzanten** [comment]
There isn't really a great way, as it would have to recursively go up. You could either add something along the lines of:

```
_and
  parent _eq X
  parent
    parent _eq X
  parent
    parent
      parent _eq X
  parent
    parent
      parent
        parent _eq X
```

or retrieve all IDs in the path first, and use an `_in` filter

**Comment by keesvanbemmel** [comment]
I think I'll just cut a corner and add a public flag to the directus_files collection. These are product images and the folder tree and amount of images will make your suggested answers not performant enough I guess.

If there are more people wanting a solution to this we can discuss here and I'll try to open a PR for it then!

Thanks for answering @rijkvanzanten

**Comment by licitdev** [comment]
We can make use of a custom extension with `action` event hooks to automatically populate the `public_access` flag within the `directus_files` collection.

## Steps
1. Create a `public_access` boolean field in `directus_folders` with the toggle interface.
2. Set the custom permissions for `directus_files` as follows:
    1. Item permissions: Add the filter of `folder > public_access > _eq > true`.
    2. Field permissions: Allow the `ID` field.
3. Set up the sample extension provided below and update the `publicFolderIDs` accordingly.

### Sample Extension
```js
module.exports = function registerHooks({ action }) {
	action('folders.create', run);
	action('folders.update', run);

	async function run(data, { database }) {
		// This field should be a boolean (eg: toggle) in directus_folders
		const publicAccessField = 'public_access';
		const publicFolderIDs = ['ab932ebc-a901-44d7-a892-53f4ffc324a3'];
		const targetFolderId = data.key ?? data.keys[0];
		const targetFolder = await getFolder(targetFolderId);

		if (!targetFolder) {
			return;
		}

		// Check if current folder is allowed public access
		let allowed = isAllowed(targetFolderId);
		let folder = targetFolder;

		// Continue checking parent chain if allowed public access
		while (!allowed && folder && folder.parent) {
			folder = await getFolder(folder.parent);

			if (!folder) {
				break;
			}

			allowed = isAllowed(folder.id);
		}

		// Update the public access into database
		if (allowed) {
			await updatePublicAccess(targetFolder.id, true);
		} else if (targetFolder[publicAccessField]) {
			await updatePublicAccess(targetFolder.id, false);
		}

		async function getFolder(id) {
			return await database.select('*').from('directus_folders').where('id', '=', id).first();
		}

		function isAllowed(id) {
			return publicFolderIDs.includes(id);
		}

		async function updatePublicAccess(id, allowed) {
			await database('directus_folders')
				.update({ [publicAccessField]: allowed })
				.where('id', '=', id);
		}
	}
};
```

## Important Notes
- This sample extension covers the updating of the `public_access` value when creating and moving of **new** folders. 
- The `public_access` value of **existing** folders have to be manually patched. Otherwise, **existing** folders can be renamed or moved which will trigger the extension to update the `public_access` value. 
- Nested folders are **not** updated when a parent folder is moved or deleted in this sample.

**Comment by baptistejouin** [comment]
Another quick implementation that you can set up is to define only two children in your folder architecture in Directus: "public" and "private." After that, create a policy for `directus_files` with a custom condition on `read` that allows every file except files included in the private folder.

```json
{
    "_and": [
        {
            "folder": {
                "_neq": "your-private-folder-uuid"
            }
        }
    ]
}
```

NOTE: This could be very useful if your use case is a deep folder architecture for public files and only a flat private folder, because you only need to refer to the UUID of this private folder. If you have subfolders in it, it's the same problem; you need to list these here.

<img width="213" alt="image" src="https://github.com/user-attachments/assets/e2f8a73c-7c90-45b5-a097-1fad196d32e7" />
<br/>
<img width="621" alt="image" src="https://github.com/user-attachments/assets/9e942308-bf58-4d00-b030-39d37e38c1b6" />
